pkg/winsvc/winsvc.go

```diff
@@ -0,0 +1,174 @@
+	CreateService(string, string, mgr.Config, ...string) (Service, error)
+	ListServices() ([]string, error)
+	OpenService(string) (Service, error)
+}
+
+type fakeServiceList struct {
+	m    *sync.Mutex
+	svcs map[string]Service
+}
+
+// write overwrites the given service to the svcs map
+func (l *fakeServiceList) write(name string, svc Service) {
+	l.m.Lock()
+	defer l.m.Unlock()
+	l.svcs[name] = svc
+}
+
+// read returns the entry with the given name, and a bool indicating if it exists or not
+func (l *fakeServiceList) read(name string) (Service, bool) {
+	l.m.Lock()
+	defer l.m.Unlock()
+	service, exists := l.svcs[name]
+	return service, exists
+}
+
+// listServiceNames returns a slice of all service names
+func (l *fakeServiceList) listServiceNames() []string {
+	l.m.Lock()
+	defer l.m.Unlock()
+	var names []string
+	for svcName, _ := range l.svcs {
+		names = append(names, svcName)
+	}
+	return names
+}
+
+// remove deletes the entry with the given name, throwing an error if it doesn't exist
+func (l *fakeServiceList) remove(name string) error {
+	l.m.Lock()
+	defer l.m.Unlock()
+	_, exists := l.svcs[name]
+	if !exists {
+		return fmt.Errorf("service does not exist")
+	}
+	delete(l.svcs, name)
+	return nil
+}
+
+func newFakeServiceList() *fakeServiceList {
+	return &fakeServiceList{
+		m:    &sync.Mutex{},
+		svcs: make(map[string]Service),
+	}
+}
+
+type testMgr struct {
+	svcList *fakeServiceList
+}
+
+// CreateService installs new service name on the system.
+// The service will be executed by running exepath binary.
+// Use config c to specify service parameters.
+// Any args will be passed as command-line arguments when
+// the service is started; these arguments are distinct from
+// the arguments passed to Service.Start or via the "Start
+// parameters" field in the service's Properties dialog box.
+func (t testMgr) CreateService(name, exepath string, config mgr.Config, args ...string) (Service, error) {
+	// Throw an error if the service already exists
+	if _, ok := t.svcList.read(name); ok {
+		return nil, fmt.Errorf("service already exists")
+	}
+	config.BinaryPathName = exepath
+	service := FakeService{
+		name:   name,
+		config: config,
+		status: svc.Status{
+			State: svc.Stopped,
+		},
+		serviceList: t.svcList,
+	}
+	t.svcList.write(name, &service)
+	return &service, nil
+}
+
+func (t testMgr) ListServices() ([]string, error) {
+	return t.svcList.listServiceNames(), nil
+}
+
+func (t testMgr) OpenService(name string) (Service, error) {
+	service, exists := t.svcList.read(name)
+	if !exists {
+		return nil, fmt.Errorf("service does not exist")
+	}
+	return service, nil
+}
+
+type Service interface {
+	Delete() error
+	Close() error
+	Start(...string) error
+	Config() (mgr.Config, error)
+	Control(svc.Cmd) (svc.Status, error)
+	Query() (svc.Status, error)
+	UpdateConfig(mgr.Config) error
+}
+
+type FakeService struct {
+	name        string
+	config      mgr.Config
+	status      svc.Status
+	serviceList *fakeServiceList
```

Comment:
This is the only thing confusing me, why does each FakeService contains a serviceList?

Reply:
They all need to share the same serviceList as the serviceList must be updated by a `FakeService.Delete()` call, across all `FakeService` structs and the `Mgr` struct. This is mocking out Windows syscalls, where there would be one service list for the whole OS.

Reply:
Thanks for the explanation. So looks like each service's `serviceList` will point to the same object (done in `NewTestMgr()`'s loop), so no need to update every mocked service's `serviceList` individually on a `Delete()` call.